[AMDGPU][MC] Disallow op_sel in some VOP3P dot instructions

jwanggit86 opened the pull request:
In v_dot4 and v_dot8 instructions with 4- or 8-bit packed data (e.g., v_dot4_u32_u8, v_dot8_u32_u4), the op_sel modifier should not be included.

jwanggit86 committed:
[AMDGPU][MC] Disallow op_sel in v_dot4 and v_dot8 with 4- or 8-bit
packed data

In v_dot4 and v_dot8 instructions with 4- or 8-bit packed data (e.g.,
v_dot4_u32_u8, v_dot8_u32_u4), the op_sel modifier should not be
allowed for GFX9 and GFX10. For GFX11 and GFX12, section 7.5.1 of
the RDNS3 ISA doc seems to indicate that OPSEL is allowed.

arsenm reviewed (commented):
I believe this is only true for gfx940, and I think the older targets did support op_sel with these. I see a note that not supporting op_sel is a change, but don't see documentation for what it did before

DadSchoorse commented:
>  but don't see documentation for what it did before

op_sel/op_sel_hi on src0/1 works the same as for other packed instructions on gfx10-11, it selects which half of the source is read. src2 op_sel is ignored.

jwanggit86 commented:
BTW, this is for issue [44257](https://github.com/llvm/llvm-project/issues/44257).

jwanggit86 commented:
> and I think the older targets did support op_sel with these.

These instructions (v_dot4, v_dot8) have fixed ways of using the 4-bit/8-bit data, so op_sel isn't really meaningful?

jwanggit86 commented:
> I believe this is only true for gfx940, and I think the older targets did support op_sel with these. I see a note that not supporting op_sel is a change, but don't see documentation for what it did before

@arsenm 
Is one of the following the note you referred to above?
```cpp
AMDGPUInstructionSelector::selectVOP3PModsImpl(
  Register Src, const MachineRegisterInfo &MRI, bool IsDOT) const {
  ...
  (void)IsDOT; // DOTs do not use OPSEL on gfx940+, check ST.hasDOTOpSelHazard()
}

class GCNSubtarget {
...
  // Cannot use op_sel with v_dot instructions.
  bool hasDOTOpSelHazard() const { return GFX940Insts || GFX11Insts; }
}
```
Even these two places are inconsistent. One says "do not use on gfx940+", another seems to apply only to gfx940 and gfx11.

Second, are you suggesting op_sel should be disallowed for gfx940+, but left alone for older targets?

arsenm commented:
> > I believe this is only true for gfx940, and I think the older targets did support op_sel with these. I see a note that not supporting op_sel is a change, but don't see documentation for what it did before
> 
> @arsenm Is one of the following the note you referred to above?

No, I'm reading the ISA documentation, not the implementation 


> 
> Even these two places are inconsistent. One says "do not use on gfx940+", another seems to apply only to gfx940 and gfx11.

This looks like it wasn't cleanly implemented to track which targets support it and which don't. I wouldn't trust whatever is implemented here, this requires double checking the hardware docs 


> 
> Second, are you suggesting op_sel should be disallowed for gfx940+, but left alone for older targets?

It should be left alone on any target where it's supported. We shouldn't just drop it everywhere

jwanggit86 commented:
> > > I believe this is only true for gfx940, and I think the older targets did support op_sel with these. I see a note that not supporting op_sel is a change, but don't see documentation for what it did before
> > 
> > 
> > @arsenm Is one of the following the note you referred to above?
> 
> No, I'm reading the ISA documentation, not the implementation
> 
> > Even these two places are inconsistent. One says "do not use on gfx940+", another seems to apply only to gfx940 and gfx11.
> 
> This looks like it wasn't cleanly implemented to track which targets support it and which don't. I wouldn't trust whatever is implemented here, this requires double checking the hardware docs
> 
> > Second, are you suggesting op_sel should be disallowed for gfx940+, but left alone for older targets?
> 
> It should be left alone on any target where it's supported. We shouldn't just drop it everywhere

@arsenm I checked the documentation. For gfx940 it is explicitly stated that op_sel is not allowed. However, it is vague (i.e., not explicitly said) for other targets. So, should we just disallow it for gfx940 for now?

jwanggit86 commented:
@arsenm ping.

jwanggit86 committed:
Allow OPSEL for GFX11 and GFX12.

Section 7.5.1 of the RDNA3 ISA doc apparently indicates that
OPSEL is allowed (although ignored) in the dot instructions
in question. Therefore, we should allow it for GFX11 and GFX12.

Sisyph commented:
Sp3 does not support op_sel for dot4* or dot8* instructions on any of gfxip 908,90a,940,1030,11, or 12. 
None of those ISA manuals say the instructions should support op_sel.
I don't see any docs on what the semantics of opsel on those instructions would be.
The only mention of opsel on these instructions in the programming guide is in GFX11 and GFX12. It says if you use inline constants they 'use 32bit inline src0/1 (ignore OPSEL)'. We use i32 fields for those operands in tablegen, so that looks fine as is.
We don't use op_sel in codegen on those instructions. So this will not be a functional change to codegen, only the asm/disasm (please correct me if I'm wrong on this).

However, @DadSchoorse  reports some semantics on how op_sel might work on these instructions. Is that claim from experimentation on hardware, or something else?

It might be best to leave the assembler/disassembler as is, and make sure we continue to not use op_sel in codegen. If we do want to change these, please change them across all targets.

arsenm commented:
> I don't see any docs on what the semantics of opsel on those instructions would be.

I don't either. I only found a note that says "is no longer supported on DOT2/4/8" for 940. That implies it was previously supported

jwanggit86 commented:
Here is a summary of my findings from the ISA docs:

- v_dot4_*, v_dot8_* use VOP3P encoding.
- MI100 ISA: Sect 12.10 "VOP3P Instructions", only 3 V_MAD_MIX* instructions use OPSEL.
- MI200 ISA: same as MI100.
- MI300 ISA: same as MI200.
- RDNA1 ISA: No v_dot4_*, v_dot8_*; 3 V_FMA_MIX* instruxtions use OPSEL.
- RDNA2 ISA: 3 V_FMA_MIX* instruxtions use OPSEL.
- RDNA3 ISA: (1) Sect 16.10 "VOP3P Instructions", 3 V_FMA_MIX* instruxtions use OPSEL.
(2) Sect 7.5.1 "Inline Constants with Packed Math", mentions "ignore OPSEL" for dot4_* and dot8_*. So apparently OPSEL
is allowed.

Based on this I'm ok with either of the following: (1) leave assembler/disassembler as is (2) disallow OPSEL for v_dot4_* and v_dot8_* for pre-GFX11 targets, as this patch does. 

@arsenm @Sisyph Please let me know your thoughts.

jwanggit86 commented:
Here's some additional info garnered from the ISA docs.

In RDNA3 doc, Sect 7.1 "Microcode Encodings" there's a statement: "OPSEL may only be used for 16-bit operands, and must be zero for any other operands/results." Since v_dot4_* and v_dot8_* don't operate on 16-bit operands, I think OPSEL is allowed but must be 0.

In RDNA2 doc, Sect 6.2.5 "OPSEL Field Restrictions" states OPSEL is only usable in opcodes listed in Table 21, which doesn't include v_dot4_* or v_dot8_. 

In MI300 doc, Sect 6.2.1 "Instruction Inputs" has a statement: "DOT instructions must not use SDWA or OPSEL.".

In MI200 doc, not explicit statements about OPSEL in Dot.

Based on the above, I'd suggest that for v_dot4_* and v_dot8_*, (1) OPSEL should be disallowed for MI300 and RDNA2 (2) OPSEL is allowed but must be 0 for RDNA3.

@arsenm @Sisyph @DadSchoorse Pls let me know your thoughts.

arsenm commented:
We should run this by hardware. If I were to guess, this is yet another case where they didn't validate it works so did not document it as supported.

rampitec commented:
I am reading SP3 pseudocode, looks like OPSEL is not used on any of the targets with this. These instructions also do not have OPSEL flag in SP3. So I guess we need to set it to zero on all targets.

jwanggit86 committed:
Allow op_sel for VOP3P DOT instructions, but value of op_sel must
be 0 if present.

jwanggit86 commented:
In the existing code, for all dot instructions for GFX940, op_sel must be 0, and op_sel_hi cannot be included. This latest commits apply the same restrictions to VOP3P dot instructions (v_dot4_*, v_dot8_*, and some v_dot2_*).

jwanggit86 committed:
Fix a problem in previous commit. For GFX940 keep the original
behavior, i.e., op_sel should be 0 for all DOT instructions,
not just the VOP3P DOTs.

jwanggit86 commented:
@arsenm @rampitec ping.

jwanggit86 commented:
@arsenm @rampitec ping.

jwanggit86 commented on llvm/test/MC/AMDGPU/dl-insts.s:
```diff
@@ -389,345 +389,23 @@ v_dot8_i32_i4 v0, v1, v2, v3
 // CHECK: encoding: [0x00,0x40,0xa3,0xd3,0x01,0x05,0x0e,0x1c]
 v_dot2_f32_f16 v0, v1, v2, v3 op_sel:[0,0]
-// CHECK: encoding: [0x00,0x50,0xa3,0xd3,0x01,0x05,0x0e,0x1c]
-v_dot2_f32_f16 v0, v1, v2, v3 op_sel:[0,1]
```
Thanks! However, `v_dot4_` and `v_dot8_` still don't support opsel pre-gfx940, right? What's the best way to differentiate between `v_dot2_` and, say `v_dot4_`?

jwanggit86 committed:
For pre-GFX940 targets, v_dot2_* allow opsel.

rampitec commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -4588,28 +4588,40 @@ bool AMDGPUAsmParser::validateOpSel(const MCInst &Inst) {
-        return false;
+  if (TSFlags & SIInstrFlags::IsDOT) {
+    // For all DOT instructions on GFX940, or VOP3P DOT instructions on all
+    // targets, i.e. v_dot2_*(except on pre-GFX940), v_dot4_* and v_dot8_*,
```
I have checked gfx10 and DOT2 supports opsel, at least for 3 instructions below...

jwanggit86 replied:
Ok, I'll look into this.

rampitec commented:
Besides asm parser there is also codegen. Can we just select VOP3Features for dot4 and dot8 instructions to include VOP3_PACKED, but not VOP3_OPSEL? So that op_sel operands will not even exist on these. I guess it will need some encoding bits to be hardcoded in their Real definitions.

jwanggit86 commented:
> Besides asm parser there is also codegen. Can we just select VOP3Features for dot4 and dot8 instructions to include VOP3_PACKED, but not VOP3_OPSEL? So that op_sel operands will not even exist on these. I guess it will need some encoding bits to be hardcoded in their Real definitions.

An earlier version of the patch did just that. Then op_sel is disallowed for v_dot4_* and v_dot8_*. The current version allows it, but it must be 0.

rampitec commented:
> @Sisyph @arsenm @rampitec So far two solutions have been tried: (1) disallow OPSEL for v_dot4_* and v_dot8_* for pre-GFX11 targets (2) allow OPSEL for all targets but its value must be 0. A 3rd option is not to make any fix and leave things as they are. Pls let me know your thoughts.

I would disallow it. Why pre-GFX11 only? It does not do anything for dot4 and dot8 on any target.

jwanggit86 commented:
> I would disallow it. Why pre-GFX11 only? It does not do anything for dot4 and dot8 on any target.

The reason for limiting this to pre-GFX11 is that Sect 7.5.1 of the RDNA3 ISA doc indicates "ignore OPSEL" for dot4_* and dot8_*, which seems to imply it's allowed. But I'm fine with disallowing it for all targets, which is actually what the 1st commit did.

rampitec commented:
> > I would disallow it. Why pre-GFX11 only? It does not do anything for dot4 and dot8 on any target.
> 
> The reason for limiting this to pre-GFX11 is that Sect 7.5.1 of the RDNA3 ISA doc indicates "ignore OPSEL" for dot4_* and dot8_*, which seems to imply it's allowed. But I'm fine with disallowing it for all targets, which is actually what the 1st commit did.

Technically you can encode it, but it does not affect the execution. No reason to handle it differently between targets, just set encoding bits to default.

jwanggit86 committed:
Revert to the 1st commit, 2753fd11e70600, which disallows
OPSEL in v_dot4_* and v_dot8_* instructions for all targets.

This reverts commit 7ae70f094f23872cbe7b5356044b93994555f562 and
all later commits.

arsenm reviewed (approved):
This matches what the documentation states. If we really want to support undocumented usage, we should make a more conscious effort to test and validate it really works

rampitec reviewed (approved):
VOP3_PACKED_NO_OPSEL is a right thing to do.
